### ISSUE TITLE
Ability/methodology of extending Geocoder parameters

### ISSUE THREAD

**Issue by KostyaEsmukov** [geopy/geopy#305]
Some geocoding services provide very rich APIs (in terms of parameters), while geopy's geocoders usually implement just a subset of it.

We need to come up with a methodology of letting people pass arbitrary parameters if they need to, probably at a cost of possible code breakage on upgrades, as underlying API versions are getting updated in minor (not major) geopy releases, as stated in [the `semver` section of the doc][semver].

Currently in this situation I'd suggest to open a PR adding the parameters you're missing. For a quick and dirty solution, you might be okay with extending the geocoder class and overloading `geocode`/`reverse` and `_call_geocoder` methods with passing extra data via `threading.local()` (but please don't do this unless you *really* have to).

Relevant discussion: https://github.com/geopy/geopy/pull/304#discussion_r195167065
Other occurrences: #222

[semver]: https://geopy.readthedocs.io/en/1.14.0/#semver

**Comment by deeplook** [comment]
I believe the only clean way of doing this would be to rely on formal API specifications like Swagger/OpenAPI files. Of course, this only shifts the risk of changing APIs for the user away from geopy to the provider of the API spec. But it also raises the question how geopy wants to detect changes in those specs? Always checking the one canonical source at runtime, at some intervals, only in case of errors? This can turn out to be an entire new mine field, and I've been partly there.

**Comment by KostyaEsmukov** [comment]
@deeplook I'm not sure if I understood your point correctly. Could you please elaborate?

> I believe the only clean way of doing **this** would be to rely on formal API specifications like Swagger/OpenAPI files.

Doing what exactly? :)

As I said before in https://github.com/geopy/geopy/pull/304#discussion_r195385063 , full support of providers' API is not a goal of geopy. For that, specialized SDK libraries for (some of the) providers exist.

I don't quite understand how tracking API changes of the providers automatically might help with simplifying geocoder classes extension.

**Comment by zeljkobekcic** [comment]
I don't know if this is still relevant but this issue is still open and I got. One possible way to achieve this is to create a dataclass (or something comparable because dataclasses require python 3.6 and above) to wrap the arguments and that provides a method like `to_query`.

A simple/minimal/incomplete example for demonstration for Nominatin could be:

```python
class NominatimParameters:

    def __init__(
            self,
            query,
            addressdetails=None,
            format=None,
            json_callback=None)

        self.query = query
        self.addressdetails=addressdetails
        self.json_callback = json_callback

    def to_query_parameters():
        query = "q=" + str(self.query)

        if self.addressdetails is not None:
            query += "&addressdetails=" + str(addressdetails)

        if self.json_callback is not None:
            query += "&json_callback=" + str(addressdetails)

        return query
```

This would make it easier to provide "full api support" (more or less because the data still needs to be parsed) without breaking existing code.

**Comment by KostyaEsmukov** [comment]
Yes, this is still relevant.

Your solution is definitely a viable one if we were designing an API from scratch, but given that `geopy` already exposes a public API, it would be hard and painful for everyone to migrate, especially given that it would give little to no benefit to those who don't miss any features in geopy.

Another problem is that some geocoders might do more than one HTTP request during a geocoding call, so the simple `to_query_parameters` method might not be enough for them (see, for example, https://github.com/geopy/geopy/blob/1b0e80dfdd4bfb0c200e84ae1553c56e7f04325a/geopy/geocoders/arcgis.py#L201)

--

The need to separate the actual HTTP requests from parameters generation is not caused by just the desire to easily add custom parameters (this problem can be mitigated rather easily by extending the geocoders or copy-pasting). The bigger challenge is providing both synchronous and asynchronous (via `await`) interface for the geocoding calls. The parameters generation part must be shared between the two, but the actual HTTP requests can't be shared, because they differ in syntax.

One possible way would be to make everything `async` by default, and create little eventloops for the synchronous calls, but it might make the geocoders' code significantly more complicated. That said, I'm still experimenting with this.

**Comment by amotl** [comment]
Hi there,

first things first: Thanks a bunch for conceiving and maintaining this convenient package.

We would like to use the Nominatim reverse geocoder with the `jsonv2` request format instead of just `json`, see also https://github.com/panodata/luftdatenpumpe/issues/22#issuecomment-578966596. Do you believe there would be a chance to make this optionally happen within `geopy` in general? If so, we will be happy to provide an appropriate pull request.

With kind regards,
Andreas.

cc @wetterfrosch

**Comment by KostyaEsmukov** [comment]
@amotl It would've been be a great feature! So yes, a PR adding an optional parameter for changing response format from `json` to `jsonv2` is certainly welcomed.

There might be a caveat though: If there are any changes in the fields which populate address and coordinates of Location, they should be handled transparently, so that this optional parameter would only affect the `raw` contents of the `Location`.

**Comment by amotl** [comment]
Dear @KostyaEsmukov,

it looks like the baseline fields (`display_name`, `address`) are all the same while `jsonv2`-format adds some additional ones. So, I believe everything should be fine.

Do you believe everything will be fine then or do you believe I am missing something here?

Thanks already and with kind regards,
Andreas.

**Comment by KostyaEsmukov** [comment]
@amotl Apologies for the late reply. I guess it should be fine, otherwise tests should tell us if it goes wrong. Also, if that would be an optional parameter, it won't break anything for anyone who won't explicitly specify the option with the new format.

**Comment by KostyaEsmukov** [comment]
I have just merged #418 to master (geopy 2.0) which allowed to add an optional asyncio support (#419).

I guess there's nothing else we can do on simplifying extending Geocoder classes without changing the public API completely. Thus I'm closing this.
